### ISSUE TITLE
Make IPFS reflink aware, dedup file storage between IPFS and user downloaded files

### ISSUE THREAD

**Issue by Jorropo** [ipfs/go-ipfs#8201]
## TL;DR

Use reflinks from the linux kernel to store IPFS's blobs for blocks and usefull files for the user (such as downloaded with `ipfs get`) using the same backing data blocks, deduping them.

## What are reflinks in too much details

IPFS's on disk datastores all needs to double store files, once in themself, once on the file system where the files is use full to you.
This is anoying and require expensive copy times *not for filestore, more on that later* and double storages.
However there is a solution ! [`copy_file_range`](https://man7.org/linux/man-pages/man2/copy_file_range.2.html) ([godoc](
https://pkg.go.dev/golang.org/x/sys/unix#CopyFileRange)) (there are various other syscalls doing more less the same thing, this one is just near perfect for this use case)

This just perform an in Kernel copy from one file descriptor to an other (with some offset and the capacity to append).
However if the file-system allows it, this doesn't make a copy, this makes a reflink (Copy-on-Write copy of the file). In practice the file isn't copied, only the entry of it in the file-system (inode) is, so both files now share the same data blocks, however unlike what a hardlink is, if anyone of thoses files is modified instead of overwriting the data of both files, a new copy is made and the modified file's inode is changed to point to it.

This allows us to store multiple copy in the file-system (such as one for the datastore and X for the user) with the space usage of one.

Some file-systems such as btrfs are even capable to build DAGs (more precisely, B+trees) for the files, so if on a 1GB file you modify 1 byte somewhere, it's probable that only the block containing this byte is copied, and so for the near 1GB unmodified part both files to still share the storage.
Appart from btrfs, other well known reflink capable file systems are XFS and ZFS.

### The modified datastore option

Like the `datastore.GCDatastore` create a `datastore.ReflinkingDatastore` that would be attempted to cast to in the construction process of IPFS.
The API could be similar to that :
```go
type ReflinkType uint
const (
  // Reflink impossible, attempting a reflink would fail or have no advantage compare to calling `Get` and copying the result
  ReflinkType_NoReflink ReflinkType = iota
  // No reflink, however a fast copy (such as a kernel one) is possible
  ReflinkType_FastCopy
  // Reflinking, so deduping the data on the storage medium
  ReflinkType_Reflink
)

type ReflinkingDatastore interface {
  Datastore

  GetReflink(key Key, file *os.File, offset uint64) error
  PutReflink(key Key, file *os.File, offset uint64, size uint64) error

  CanReflink(file *os.File) ReflinkType
}
```
About `GetReflink` and `PutReflink` :
The `*os.File` is just to fetch the file descriptor, could be replaced by an `uintptr` maybe, but I like `*os.File` more as this makes the type of the API explicit.
`offset` and `size` are used because from one single file you could create multiple blocks, or append multiple blocks into one file.
`offset` is the offset into the `file`.
`size` is the size of the copy, on `GetReflink` this is implied by the size of the key.
Implementations must not `seek` `file` as this could be called concurrently.

It should be obvious how flat-fs could be made compatible with this using `copy_file_range`.

`CanReflink` test the file and returns the appropriate value (see the comments in the enum).

Then while getting a file, we would first test if we are from the same machine as the API caller (as obviously nothing can be done if we are not in the same file-system).
If so, open the target file, download it, and walk the block graph.
Test if it make sense to use the reflink API (instead fallback to the current implementation).
While walking the graph, for each block just call `GetReflink` passing the correct offsets.

The reverse applies while adding (just send the correct `size` + `offset`).

### Cross file system

Reflinks doesn't work cross file-systems (*most of the time*, btrfs can reflink from an ext2,3,4 partition if it is readonly, this is used when converting ext partitions into btrfs, we can ignore this).
So to not fallback to a copy across multiple file-systems we would need to support some kind of multiple datastore system (or a datastore capable to store in multiple paths), allowing to have a datastore folder in each file-system and picking the one reflink compatible.

The main issue is how we would handle data managment.
Let's assume something quite simple, we have a main datastore and multiple flat-fs (one for each file system), the main datastore stores every block belonging no where and the flat-fs index for one stored in a an other file systems.
So while getting a file, you would look up in the main datastore to see if the file is already in the flat-fs of the destination file-system.
If it is then just reflink on top of the target file as usual and continue for other blocks.
If it is not then first copy it in the flat-fs and then reflink it and continue.
However if the user then remove or fully modify the file he just downloaded, we are keeping a reference to it for nothing, preventing to free up space.
Idealy we would need some kind of purgeable reflink that the OS could remove to free up space if needed. But I don't think linux has this kind of thing yet.

## The main drawbacks

- This is a relatively new feature in the linux kernel. The API for `copy_file_range` is only stable since `5.3` (it has been introduced much earlier but this was a buggy inconsistent API, and it's generally advised to avoid it prior to `5.3`), this is debian 11 (not even yet marked stable) or ubuntu 19.10 (20.04 for LTS) for example.
  It exists other syscalls less suited to the task but that could still make it work tho, I think this is not an issue, at first we can use `uname` to enable it depending on `x >= 5.3` while we figure out the architecture and then, if someone with an old kernel wants it, they could try to make it work with older kernels.
- This kind of aggressive multiple reflinking will lead to atrocious fragmentation.
  We can expect the performance of this feature to be really bad on hdd, not really while copying as file-systems with allocation group are capable to deal with fragmented writes quite well. But while reading too.
  I don't think this is a reason to not do it, it's not because HDDs aren't suitable to the task that SSDs should be holded back like that.
- This makes space usage reports confusing and the disk full situation more problematic.
  Thoses are general tradeoff of reflinking file-systems, someone running them should already know and expect that.

## About filestore

This is in some way better than filestore as the original files added through reflink would be modifiable, movable, ... without clobering IPFS's datastore.

**Comment by aschmahmann** [comment]
👍 Taking advantage of copy-on-write support to create a better Filestore makes sense to me (this was also suggested in https://github.com/ipfs/go-ipfs/issues/7557).

I tend to think that support for reflinks should be something the user explicitly opts into (e.g. via config flag) otherwise docs become basically impossible and you end up with something like "you can save space by using the Filestore on Linux and you're free to change your existing files, but on Windows everything will break if you modify your files".

As mentioned in the above issue deduping storage during `ipfs get` is going to require more plumbing changes than `ipfs add` https://github.com/ipfs/go-ipfs/issues/3981, just making a new datastore (even one the replaces the filestore) is insufficient.

Overall copy-on-write is generally great for data that is mostly immutable, looking forward to hearing how your experimentation goes.

**Comment by Jorropo** [comment]
Update: I have created a Go proposal which would make implementing this far easier: golang/go#52383

**Comment by calumapplepie** [comment]
Be careful not to overthink it; I for one would be fine with the system temporarily using more disk space if it means long-term savings.  If we know where all the files are in the datastore and file system, can we not simply load them in normally and then, later and separately, use the IOCTLs a la jdupes to remove the duplicates?

**Comment by Jorropo** [comment]
@calumapplepie that sounds harder to me than doing it right.

The biggest issue here is that our datastore returns `[]byte`.
If we could make it return `io.Reader` then we can use type assertion like `io.Copy` is doing and call `io.WriteTo` or `io.ReadFrom` after which go takes care of deduping for us because `*os.File` implements `io.ReaderFrom` by using `copy_file_range`.

**Comment by maybeetree** [comment]
Hello,

Is there any update on this? I would like to use a private IPFS swarm as a sort of backup/archival solution,
and it would be really nice to have this feature.

**Comment by karawitan** [comment]
## Proposal: Reflink-aware `ipfs get` via a `Reflinker` blockstore capability

This is a concrete implementation plan for #8201 (and the older #7557), updated for the current state of kubo (v0.44, post-badger-removal, raw-leaves default, flatfs recommended).

### What changed since this issue was opened

Several things make this more tractable now than in 2021:

1. **Raw leaves are the default.** Since IPIP-0499, `ipfs add` produces raw-leaf CIDs by default. A raw leaf block's bytes *are* the file chunk — no UnixFS protobuf wrapper. This means the block file in flatfs contains exactly the bytes that belong in the output file at that offset. Reflinking the whole block file into the right position in the output file is a 1:1 byte mapping, no offset arithmetic into the block needed.

2. **flatfs is the recommended datastore** (badgerds removed in #11455). flatfs stores each block as an individual file on disk — the exact structure `FICLONERANGE` needs. Other datastores (levelds, pebbleds) store blocks inside a single database file and cannot reflink individual blocks; that's fine, they just don't get this feature.

3. **`golang.org/x/sys/unix`** ships `IoctlFileCloneRange` and `CopyFileRange`. No CGo, no manual syscall plumbing.

4. **Reflink-capable filesystems are mainstream**: btrfs, XFS (reflink=1 is the default), bcachefs. The kernel `copy_file_range(2)` API has been stable since 5.3 (2021); every supported distro kernel has it.

### Why this is not a protocol change

Reflinks are a **local OS-level operation** between two files on the same filesystem. No wire format changes, no on-disk datastore format changes, no peer-visible behavior. The blockstore still stores the same blocks with the same keys. The output file is a local file the operator asked for. This is a local I/O optimization, not an IPIP.

### Architecture: where each piece goes

Per the boxo/kubo split (AGENTS.md): the reusable capability goes in boxo, the flatfs implementation in go-ds-flatfs, the CLI surface and wiring in kubo.

```
┌─────────────────────────────────────────────────────┐
│ kubo: core/commands/get.go                           │
│   --reflink flag                                     │
│   type-asserts files.Node → boxo Reflinker           │
│   falls back to tar extraction if unsupported         │
└──────────────────────┬──────────────────────────────┘
                       │ uses
┌──────────────────────▼──────────────────────────────┐
│ boxo: ipld/unixfs/io                                 │
│   Reflinker interface on DagReader                   │
│   walks leaf blocks, calls blockstore.Reflink        │
│   for each, at the correct offset in dest file       │
└──────────────────────┬──────────────────────────────┘
                       │ calls
┌──────────────────────▼──────────────────────────────┐
│ boxo: blockstore                                     │
│   Reflinker interface (cid → dest *os.File, offset)  │
│   blockstore wrapper passes through to datastore     │
└──────────────────────┬──────────────────────────────┘
                       │ delegates
┌──────────────────────▼──────────────────────────────┐
│ go-datastore: Reflinker interface                    │
│   Reflink(ctx, key, dest *os.File, offset int64)     │
└──────────────────────┬──────────────────────────────┘
                       │ implements
┌──────────────────────▼──────────────────────────────┐
│ go-ds-flatfs: flatfs.Datastore                       │
│   opens block file, FICLONERANGE → dest at offset    │
└─────────────────────────────────────────────────────┘
```

### Interface design

**go-datastore** — one new optional interface:

```go
// Reflinker is implemented by datastores that can copy-on-write reflink
// a value into a destination file, sharing the underlying storage blocks.
type Reflinker interface {
    // Reflink creates a copy-on-write clone of the value stored under key
    // into dest at the given offset. Returns ErrReflinkUnsupported if the
    // datastore or filesystem cannot reflink.
    Reflink(ctx context.Context, key Key, dest *os.File, offset int64) error
}

var ErrReflinkUnsupported = errors.New("datastore does not support reflink")
```

**go-ds-flatfs** — implements `Reflinker`:

```go
func (fs *Datastore) Reflink(ctx context.Context, key ds.Key, dest *os.File, offset int64) error {
    path := fs.encode(key)           // existing path computation
    src, err := os.OpenFile(path, os.O_RDONLY, 0)
    if err != nil { return err }
    defer src.Close()

    fi, err := src.Stat()
    if err != nil { return err }

    // FICLONERANGE: clone src[0..size] into dest[offset..offset+size]
    return unix.IoctlFileCloneRange(dest.Fd(), &unix.FileCloneRange{
        SrcFd:      src.Fd(),
        SrcOffset:  0,
        SrcLength:  uint64(fi.Size()),
        DestOffset: uint64(offset),
    })
    // If FICLONERANGE fails (EXDEV, EOPNOTSUPP, ENOTSUP), return
    // ErrReflinkUnsupported so the caller can fall back to a copy.
}
```

**boxo blockstore** — exposes `Reflinker` through the blockstore layers:

```go
// Reflinker is implemented by blockstores backed by a reflink-capable datastore.
type Reflinker interface {
    // Reflink creates a copy-on-write clone of the block identified by cid
    // into dest at the given offset.
    Reflink(ctx context.Context, c cid.Cid, dest *os.File, offset int64) error
}
```

The `blockstore.NewBlockstore` wrapper type-asserts its datastore to `ds.Reflinker` and proxies through. Intermediate wrappers (CachedBlockstore, IdStore, VerifBS) pass it through via interface assertion, the same way `PersistentDatastore` / `GCDatastore` propagate today.

**boxo ipld/unixfs/io** — `DagReader` gains an optional `Reflinker` method:

```go
// Reflinker is implemented by DagReaders that can reflink their leaf blocks
// directly into a destination file, sharing storage with the blockstore.
type Reflinker interface {
    // ReflinkTo writes the file's content to dest by copy-on-write reflinking
    // each leaf block from the blockstore into dest at the correct offset.
    // Returns ErrReflinkUnsupported if the blockstore or filesystem can't reflink.
    ReflinkTo(ctx context.Context, dest *os.File) error
}
```

The implementation walks the UnixFS DAG (the same leaf traversal `Read` already does), and for each raw leaf block calls `blockstore.Reflinker.Reflink(ctx, leafCID, dest, currentOffset)`. For non-raw-leaf blocks (legacy UnixFS protobuf nodes), it falls back to `ErrReflinkUnsupported` — the protobuf wrapping means the block file bytes aren't a 1:1 match for the output file region.

**kubo core/commands/get.go** — new `--reflink` flag:

```
ipfs get <path> --reflink[=auto|always|never]
```

- `never` (default): current behavior (tar extraction, byte copy)
- `auto`: if the blockstore implements `Reflinker` and the output path is on the same filesystem as the datastore, use reflinks; otherwise fall back silently
- `always`: use reflinks, error if unsupported (no silent fallback)

The `PostRun: CLI` handler checks if the `files.Node` from `api.Unixfs().Get()` implements `boxo.UnixFSReflinker`, and if `--reflink` is set, calls `ReflinkTo` instead of the tar-based `writeExtracted` path. For directories, it walks the `files.Node` tree and reflinks each file individually (bypassing tar entirely for the reflink path).

### Implementation phases

**Phase 1 — `ipfs get --reflink` (read path, flatfs only)**

This is the highest-value, lowest-risk slice:
- go-datastore: add `Reflinker` interface + `ErrReflinkUnsupported`
- go-ds-flatfs: implement `Reflinker` via `FICLONERANGE`
- boxo: add `blockstore.Reflinker`, propagate through wrappers; add `ReflinkTo` on `DagReader`
- kubo: add `--reflink` flag to `get` command, wire up, add `docs/config.md`-equivalent CLI docs
- Tests: unit test the flatfs reflink (on btrfs/XFS in CI), integration test `ipfs get --reflink` produces a file with the same content and shared extents (`filefrag -e` or `FIEMAP`)

**Phase 2 — `ipfs add --reflink` (write path)**

Symmetric: when adding with flatfs + raw leaves, reflink chunks from the source file into block files instead of writing bytes. The chunker still reads the data to hash it, but the block write becomes a reflink:
- go-datastore: add `PutReflink(ctx, key, src *os.File, offset, size int64)` to `Reflinker`
- go-ds-flatfs: implement `PutReflink` (create block file, `FICLONERANGE` from src)
- boxo: wire into the importer/builder when raw leaves are used
- kubo: add `--reflink` to `ipfs add`

**Phase 3 — macOS APFS support (optional)**

APFS supports `clonefile(2)` / `FICLONERANGE` equivalent (`clonefileat`). The `Reflinker` interface is platform-agnostic; only the flatfs implementation needs a platform-specific path. Gate with build tags.

### Limitations (documented in `--help`)

1. **flatfs only.** levelds, pebbleds store blocks in a single database file — no per-block file to reflink from. `--reflink=auto` silently falls back to copy.

2. **Raw-leaf files only.** Legacy UnixFS (protobuf-wrapped leaves) can't be reflinked because the block file contains the protobuf, not just the data. `--reflink=auto` falls back; `--reflink=always` errors with a clear message.

3. **Same filesystem.** `FICLONERANGE` returns `EXDEV` across filesystem boundaries. `--reflink=auto` falls back to copy. (The issue's cross-filesystem multi-datastore idea is explicitly out of scope for this proposal — it requires a multi-path datastore that doesn't exist.)

4. **Linux only (Phase 1-2).** `FICLONERANGE` is a Linux ioctl. macOS APFS support is Phase 3.

5. **Fragmentation.** Reflinking many small blocks into one file creates a fragmented extent map. Fine on SSDs, potentially slow on HDDs. The operator opts in with `--reflink`.

6. **No GC interaction.** CoW means the output file and the blockstore blocks are independent after reflink. GC can remove the block; the output file's reflinked region is copied-on-write and stays. Deleting the output file doesn't touch the blockstore. Clean separation, no new GC complexity.

### Relationship to filestore and #10016

- **filestore** (`--nocopy` / `ipfs add --nocopy`): stores a *reference* to the original file; reads blocks from the original file on demand. Fragile — if the original file moves or changes, blocks are lost. Reflinks are the robust alternative: the data exists in both places but shares physical storage via CoW. Both can coexist; `--reflink` is for users who want the blockstore to own the data but also want a local copy without doubling storage.
- **#10016** (closed): proposed changing the block storage format entirely (store whole files, metadata for chunks). That's a much bigger change to on-disk format. This proposal achieves the same dedup goal without changing how blocks are stored — it changes how the *output* file is written.

**Comment by karawitan** [comment]
## Proposal: Reflink-aware `ipfs get` via a `Reflinker` blockstore capability

This is a concrete implementation plan for #8201 (and the older #7557), updated for the current state of kubo (v0.44, post-badger-removal, raw-leaves default, flatfs recommended).

### What changed since this issue was opened

Several things make this more tractable now than in 2021:

1. **Raw leaves are the default.** Since IPIP-0499, `ipfs add` produces raw-leaf CIDs by default. A raw leaf block's bytes *are* the file chunk — no UnixFS protobuf wrapper. This means the block file in flatfs contains exactly the bytes that belong in the output file at that offset. Reflinking the whole block file into the right position in the output file is a 1:1 byte mapping, no offset arithmetic into the block needed.

2. **flatfs is the recommended datastore** (badgerds removed in #11455). flatfs stores each block as an individual file on disk — the exact structure `FICLONERANGE` needs. Other datastores (levelds, pebbleds) store blocks inside a single database file and cannot reflink individual blocks; that's fine, they just don't get this feature.

3. **`golang.org/x/sys/unix`** ships `IoctlFileCloneRange` and `CopyFileRange`. No CGo, no manual syscall plumbing.

4. **Reflink-capable filesystems are mainstream**: btrfs, XFS (reflink=1 is the default), bcachefs. The kernel `copy_file_range(2)` API has been stable since 5.3 (2021); every supported distro kernel has it.

### Why this is not a protocol change

Reflinks are a **local OS-level operation** between two files on the same filesystem. No wire format changes, no on-disk datastore format changes, no peer-visible behavior. The blockstore still stores the same blocks with the same keys. The output file is a local file the operator asked for. This is a local I/O optimization, not an IPIP.

### Architecture: where each piece goes

Per the boxo/kubo split (AGENTS.md): the reusable capability goes in boxo, the flatfs implementation in go-ds-flatfs, the CLI surface and wiring in kubo.

```
┌─────────────────────────────────────────────────────┐
│ kubo: core/commands/get.go                           │
│   --reflink flag                                     │
│   type-asserts files.Node → boxo Reflinker           │
│   falls back to tar extraction if unsupported         │
└──────────────────────┬──────────────────────────────┘
                       │ uses
┌──────────────────────▼──────────────────────────────┐
│ boxo: ipld/unixfs/io                                 │
│   Reflinker interface on DagReader                   │
│   walks leaf blocks, calls blockstore.Reflink        │
│   for each, at the correct offset in dest file       │
└──────────────────────┬──────────────────────────────┘
                       │ calls
┌──────────────────────▼──────────────────────────────┐
│ boxo: blockstore                                     │
│   Reflinker interface (cid → dest *os.File, offset)  │
│   blockstore wrapper passes through to datastore     │
└──────────────────────┬──────────────────────────────┘
                       │ delegates
┌──────────────────────▼──────────────────────────────┐
│ go-datastore: Reflinker interface                    │
│   Reflink(ctx, key, dest *os.File, offset int64)     │
└──────────────────────┬──────────────────────────────┘
                       │ implements
┌──────────────────────▼──────────────────────────────┐
│ go-ds-flatfs: flatfs.Datastore                       │
│   opens block file, FICLONERANGE → dest at offset    │
└─────────────────────────────────────────────────────┘
```

### Interface design

**go-datastore** — one new optional interface:

```go
// Reflinker is implemented by datastores that can copy-on-write reflink
// a value into a destination file, sharing the underlying storage blocks.
type Reflinker interface {
    // Reflink creates a copy-on-write clone of the value stored under key
    // into dest at the given offset. Returns ErrReflinkUnsupported if the
    // datastore or filesystem cannot reflink.
    Reflink(ctx context.Context, key Key, dest *os.File, offset int64) error
}

var ErrReflinkUnsupported = errors.New("datastore does not support reflink")
```

**go-ds-flatfs** — implements `Reflinker`:

```go
func (fs *Datastore) Reflink(ctx context.Context, key ds.Key, dest *os.File, offset int64) error {
    path := fs.encode(key)           // existing path computation
    src, err := os.OpenFile(path, os.O_RDONLY, 0)
    if err != nil { return err }
    defer src.Close()

    fi, err := src.Stat()
    if err != nil { return err }

    // FICLONERANGE: clone src[0..size] into dest[offset..offset+size]
    return unix.IoctlFileCloneRange(dest.Fd(), &unix.FileCloneRange{
        SrcFd:      src.Fd(),
        SrcOffset:  0,
        SrcLength:  uint64(fi.Size()),
        DestOffset: uint64(offset),
    })
    // If FICLONERANGE fails (EXDEV, EOPNOTSUPP, ENOTSUP), return
    // ErrReflinkUnsupported so the caller can fall back to a copy.
}
```

**boxo blockstore** — exposes `Reflinker` through the blockstore layers:

```go
// Reflinker is implemented by blockstores backed by a reflink-capable datastore.
type Reflinker interface {
    // Reflink creates a copy-on-write clone of the block identified by cid
    // into dest at the given offset.
    Reflink(ctx context.Context, c cid.Cid, dest *os.File, offset int64) error
}
```

The `blockstore.NewBlockstore` wrapper type-asserts its datastore to `ds.Reflinker` and proxies through. Intermediate wrappers (CachedBlockstore, IdStore, VerifBS) pass it through via interface assertion, the same way `PersistentDatastore` / `GCDatastore` propagate today.

**boxo ipld/unixfs/io** — `DagReader` gains an optional `Reflinker` method:

```go
// Reflinker is implemented by DagReaders that can reflink their leaf blocks
// directly into a destination file, sharing storage with the blockstore.
type Reflinker interface {
    // ReflinkTo writes the file's content to dest by copy-on-write reflinking
    // each leaf block from the blockstore into dest at the correct offset.
    // Returns ErrReflinkUnsupported if the blockstore or filesystem can't reflink.
    ReflinkTo(ctx context.Context, dest *os.File) error
}
```

The implementation walks the UnixFS DAG (the same leaf traversal `Read` already does), and for each raw leaf block calls `blockstore.Reflinker.Reflink(ctx, leafCID, dest, currentOffset)`. For non-raw-leaf blocks (legacy UnixFS protobuf nodes), it falls back to `ErrReflinkUnsupported` — the protobuf wrapping means the block file bytes aren't a 1:1 match for the output file region.

**kubo core/commands/get.go** — new `--reflink` flag:

```
ipfs get <path> --reflink[=auto|always|never]
```

- `never` (default): current behavior (tar extraction, byte copy)
- `auto`: if the blockstore implements `Reflinker` and the output path is on the same filesystem as the datastore, use reflinks; otherwise fall back silently
- `always`: use reflinks, error if unsupported (no silent fallback)

The `PostRun: CLI` handler checks if the `files.Node` from `api.Unixfs().Get()` implements `boxo.UnixFSReflinker`, and if `--reflink` is set, calls `ReflinkTo` instead of the tar-based `writeExtracted` path. For directories, it walks the `files.Node` tree and reflinks each file individually (bypassing tar entirely for the reflink path).

### Implementation phases

**Phase 1 — `ipfs get --reflink` (read path, flatfs only)**

This is the highest-value, lowest-risk slice:
- go-datastore: add `Reflinker` interface + `ErrReflinkUnsupported`
- go-ds-flatfs: implement `Reflinker` via `FICLONERANGE`
- boxo: add `blockstore.Reflinker`, propagate through wrappers; add `ReflinkTo` on `DagReader`
- kubo: add `--reflink` flag to `get` command, wire up, add `docs/config.md`-equivalent CLI docs
- Tests: unit test the flatfs reflink (on btrfs/XFS in CI), integration test `ipfs get --reflink` produces a file with the same content and shared extents (`filefrag -e` or `FIEMAP`)

**Phase 2 — `ipfs add --reflink` (write path)**

Symmetric: when adding with flatfs + raw leaves, reflink chunks from the source file into block files instead of writing bytes. The chunker still reads the data to hash it, but the block write becomes a reflink:
- go-datastore: add `PutReflink(ctx, key, src *os.File, offset, size int64)` to `Reflinker`
- go-ds-flatfs: implement `PutReflink` (create block file, `FICLONERANGE` from src)
- boxo: wire into the importer/builder when raw leaves are used
- kubo: add `--reflink` to `ipfs add`

**Phase 3 — macOS APFS support (optional, limited)**

APFS supports `clonefile(2)` / `clonefileat(2)` / `fclonefileat(2)` (macOS 10.12+, APFS only). However, unlike Linux's `FICLONERANGE`, the Darwin `clonefile` family is **full-file only** — there is no byte-range variant. This is a fundamental constraint:

- **`ipfs get --reflink`**: Linux reflinks each leaf block file into the correct offset of the output file via `FICLONERANGE`. Darwin cannot place a block's bytes at an arbitrary offset inside a larger destination file. The Darwin path would need to fall back to `pwrite` (losing CoW sharing), or only use `clonefile` when the output is a single-block file (rare for real files).
- **`ipfs add --reflink`**: Linux reflinks a byte-range from the source file into a new block file. Darwin cannot clone a sub-range — it would need to `pread`+`pwrite` the chunk (no CoW), or clone the whole source file and truncate (wasteful, only works for single-block files).

The `Reflinker` interface is platform-agnostic; the flatfs implementation needs a platform-specific path:
- **Linux** (`linux` build tag): `FICLONERANGE` via `unix.IoctlFileCloneRange` — full range support.
- **Darwin** (`darwin` build tag): `clonefile`/`fclonefileat` — full-file only; `Reflink` returns `ErrReflinkUnsupported` for range requests, falling back to copy in `--reflink=auto` mode. Only single-block files (where the block *is* the entire file) can use true CoW on Darwin.

This means macOS APFS support is **partial**: the CoW dedup benefit is only available for files that fit in a single UnixFS block (typically ≤256KB with default chunker). Multi-block files fall back to a regular copy on Darwin.

### Limitations (documented in `--help`)

1. **flatfs only.** levelds, pebbleds store blocks in a single database file — no per-block file to reflink from. `--reflink=auto` silently falls back to copy.

2. **Raw-leaf files only.** Legacy UnixFS (protobuf-wrapped leaves) can't be reflinked because the block file contains the protobuf, not just the data. `--reflink=auto` falls back; `--reflink=always` errors with a clear message.

3. **Same filesystem.** `FICLONERANGE` returns `EXDEV` across filesystem boundaries. `--reflink=auto` falls back to copy. (The issue's cross-filesystem multi-datastore idea is explicitly out of scope for this proposal — it requires a multi-path datastore that doesn't exist.)

4. **Linux only (Phase 1-2).** `FICLONERANGE` is a Linux ioctl. macOS APFS support is Phase 3, but is **partial** — Darwin's `clonefile(2)` is full-file-only with no byte-range variant, so multi-block files fall back to a regular copy. Only single-block files (≤256KB with default chunker) get true CoW on macOS.

5. **Fragmentation.** Reflinking many small blocks into one file creates a fragmented extent map. Fine on SSDs, potentially slow on HDDs. The operator opts in with `--reflink`.

6. **No GC interaction.** CoW means the output file and the blockstore blocks are independent after reflink. GC can remove the block; the output file's reflinked region is copied-on-write and stays. Deleting the output file doesn't touch the blockstore. Clean separation, no new GC complexity.

### Relationship to filestore and #10016

- **filestore** (`--nocopy` / `ipfs add --nocopy`): stores a *reference* to the original file; reads blocks from the original file on demand. Fragile — if the original file moves or changes, blocks are lost. Reflinks are the robust alternative: the data exists in both places but shares physical storage via CoW. Both can coexist; `--reflink` is for users who want the blockstore to own the data but also want a local copy without doubling storage.
- **#10016** (closed): proposed changing the block storage format entirely (store whole files, metadata for chunks). That's a much bigger change to on-disk format. This proposal achieves the same dedup goal without changing how blocks are stored — it changes how the *output* file is written.
